kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/KafkaProxyFrontendHandler.java

```diff
@@ -553,4 +588,53 @@ public void channelActive(ChannelHandlerContext ctx) throws Exception {
     public String toString() {
         return "KafkaProxyFrontendHandler{inbound = " + inboundCtx.channel() + ", state = " + state + "}";
     }
+
+    private void outboundChannelUsable() {
+        if (state != State.CONNECTED && state != State.NEGOTIATING_TLS) {
+            throw illegalState(null);
+        }
+        LOGGER.trace("{}: outboundChannelUsable: {}", inboundCtx.channel().id(), outboundCtx.channel().id());
+        // connection is complete, so first forward the buffered message
+        for (Object bufferedMsg : bufferedMsgs) {
+            forwardOutbound(outboundCtx, bufferedMsg);
+        }
+        bufferedMsgs = null; // don't pin in memory once we no longer need it
+        if (pendingReadComplete) {
+            pendingReadComplete = false;
+            channelReadComplete(outboundCtx);
+        }
+        state = State.OUTBOUND_ACTIVE;
+
+        var inboundChannel = this.inboundCtx.channel();
+        // once buffered message has been forwarded we enable auto-read to start accepting further messages
+        inboundChannel.config().setAutoRead(true);
+    }
+
+    private void onUpstreamSslContext(SslContext sslContext) {
+        this.state = State.NEGOTIATING_TLS;
+        final HostPort remote = (HostPort) outboundCtx.channel().attr(UPSTREAM_PEER_KEY).get();
+        final SslHandler sslHandler = this.outboundCtx.pipeline().get(SslHandler.class);
+        sslHandler.handshakeFuture().addListener(handshakeFuture -> {
+            if (handshakeFuture.isSuccess()) {
+                this.outboundChannelUsable();
+            }
+            else {
+                this.onConnectionFailed(remote, handshakeFuture, this.inboundCtx.channel());
+            }
+        });
+    }
+
+    private Optional<?> onSslHandshakeException(Throwable throwable) {
+        Object result;
+        final Object triggerMsg = bufferedMsgs != null ? bufferedMsgs.get(0) : null;
+        LOGGER.info("{}: unable to complete TLS negotiation with {}", inboundCtx.channel().id(), outboundCtx.channel().attr(UPSTREAM_PEER_KEY).get());
```

Comment:
The admin still not going to be able to see the 'certificate expired' message in the logs.
Could you chain the thowable?
I'd be happy with your `.setCause(LOGGER.isDebugEnabled() ? throwable : null)` and `Increase log level to DEBUG for stacktrace` pattern you took above.

Reply:
yeah thats a good point. Have fixed in 5685c69